pkg/controller/namespace/deletion/namespaced_resources_deleter.go

```diff
@@ -526,6 +528,38 @@ func (d *namespacedResourcesDeleter) deleteAllContent(ctx context.Context, ns *v
 		gvrToNumRemaining:        map[schema.GroupVersionResource]int{},
 		finalizersToNumRemaining: map[string]int{},
 	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.OrderedNamespaceDeletion) {
+		// TODO: remove this log when the feature gate is enabled by default
+		logger.V(5).Info("Namespace controller - OrderedNamespaceDeletion feature gate is enabled", "namespace", namespace)
+		// Ensure all pods in the namespace are deleted first
+		podsGVR := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "pods"}
+		gvrDeletionMetadata, err := d.deleteAllContentForGroupVersionResource(ctx, podsGVR, namespace, namespaceDeletedAt)
```

Comment:
Somewhere we discussed whether we need to wait for actual delete or just that all pods are "deleting" and have been stopped.  That would require actually looking into the Pod object itself and knowing about the kubelet state.

I think this (waiting for deleted) is fine for now, but do you think we should consider the above?

Reply:
In current behavior, the pod are deleting and have been stopped would prevent the pod from serving traffic. I am not quite sure about the necessity of the use case of "must actual delete pod before deleting other resources". Is there a concern of "deleting pod" be exposure to security risk?

Reply:
Users can't permanently stop a pod from being stopped, but they can stop it from being deleted (via finalizers).

It may not matter, since the NS deletion is "stuck" either way.

And we don't have a general way to know that an arbitrary object is "stopped" (it's pod-specific), much less "stopped and guaranteeed not to restart".

OK, you talked me out of it.